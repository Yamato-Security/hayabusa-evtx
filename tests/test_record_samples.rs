mod fixtures;
use fixtures::*;

use evtx::{EvtxParser, ParserSettings};
use pretty_assertions::assert_eq;

#[test]
fn test_event_xml_sample() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/security.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    assert_eq!(
        first_record.data.lines().map(str::trim).collect::<String>(),
        include_str!("../samples/security_event_1.xml")
            .lines()
            .map(str::trim)
            .collect::<String>()
    );
}

#[test]
fn test_event_json_sample() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/security.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    assert_eq!(
        first_record.data.lines().map(str::trim).collect::<String>(),
        include_str!("../samples/security_event_1.json")
            .lines()
            .map(str::trim)
            .collect::<String>()
    );
}

#[test]
fn test_event_json_sample_with_event_data() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/2-system-Security-dirty.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    assert_eq!(
        first_record.data.lines().map(str::trim).collect::<String>(),
        include_str!("../samples/event_with_eventdata.json")
            .lines()
            .map(str::trim)
            .collect::<String>()
    );
}

#[test]
fn test_event_xml_sample_with_event_data() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/2-system-Security-dirty.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    assert_eq!(
        first_record.data.lines().map(str::trim).collect::<String>(),
        include_str!("../samples/event_with_eventdata.xml")
            .lines()
            .map(str::trim)
            .collect::<String>()
    );
}

#[test]
fn test_event_json_sample_with_event_data_with_attributes_and_text() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/system.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    assert_eq!(
        first_record.data.lines().map(str::trim).collect::<String>(),
        include_str!("../samples/event_with_text_and_attributes.json")
            .lines()
            .map(str::trim)
            .collect::<String>()
    );
}

#[test]
fn test_event_xml_sample_with_event_data_with_attributes_and_text() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/system.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    assert_eq!(
        first_record.data.lines().map(str::trim).collect::<String>(),
        include_str!("../samples/event_with_text_and_attributes.xml")
            .lines()
            .map(str::trim)
            .collect::<String>()
    );
}

#[test]
fn test_event_xml_sample_with_user_data() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    println!("{}", first_record.data);

    assert_eq!(
        first_record.data.lines().map(str::trim).collect::<String>(),
        include_str!("../samples/event_with_template_as_substitution.xml")
            .lines()
            .map(str::trim)
            .collect::<String>()
    );
}

#[test]
fn test_event_xml_sample_with_entity_ref() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let record = parser
        .records()
        .filter_map(|record| record.ok())
        .find(|record| record.event_record_id == 28)
        .expect("record to parse correctly");

    println!("{}", record.data);

    assert_eq!(
        record.data.lines().map(str::trim).collect::<String>(),
        include_str!("../samples/event_with_entity_ref.xml")
            .lines()
            .map(str::trim)
            .collect::<String>()
    );
}

#[test]
fn test_event_xml_sample_with_entity_ref_2() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let record = parser
        .records()
        .filter_map(|record| record.ok())
        .find(|record| record.event_record_id == 25)
        .expect("record to parse correctly");

    println!("{}", record.data);

    assert_eq!(
        record.data.lines().map(str::trim).collect::<String>(),
        include_str!("../samples/event_with_entity_ref_2.xml")
            .lines()
            .map(str::trim)
            .collect::<String>()
    );
}

#[test]
fn test_event_json_with_multiple_nodes_same_name() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let record = parser
        .records_json()
        .filter_map(|record| record.ok())
        .find(|record| record.event_record_id == 28)
        .expect("record to parse correctly");

    println!("{}", record.data);

    assert_eq!(
        record.data.lines().map(str::trim).collect::<String>(),
        include_str!("../samples/event_with_multiple_nodes_same_name.json")
            .lines()
            .map(str::trim)
            .collect::<String>()
    );
}

#[test]
fn test_event_json_sample_with_separate_json_attributes() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/Application.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(
            ParserSettings::new()
                .num_threads(1)
                .separate_json_attributes(true),
        );

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    assert_eq!(
        first_record.data.lines().map(str::trim).collect::<String>(),
        include_str!("../samples/application_event_1_separate_attributes.json")
            .lines()
            .map(str::trim)
            .collect::<String>()
    );
}
