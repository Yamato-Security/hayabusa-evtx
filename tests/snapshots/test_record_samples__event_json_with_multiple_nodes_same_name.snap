---
source: tests/test_record_samples.rs
expression: "&value"
---
{
  "Event": {
    "#attributes": {
      "xmlns": "http://schemas.microsoft.com/win/2004/08/events/event"
    },
    "System": {
      "Provider": {
        "#attributes": {
          "Name": "Microsoft-Windows-CAPI2",
          "Guid": "{5bbca4a8-b209-48dc-a8c7-b23d3e5216fb}"
        }
      },
      "EventID": 53,
      "Version": 0,
      "Level": 4,
      "Task": 53,
      "Opcode": 2,
      "Keywords": "0x4000000000000036",
      "TimeCreated": {
        "#attributes": {
          "SystemTime": "2017-05-19T02:02:36.203125Z"
        }
      },
      "EventRecordID": 28,
      "Correlation": null,
      "Execution": {
        "#attributes": {
          "ProcessID": 1396,
          "ThreadID": 2132
        }
      },
      "Channel": "Microsoft-Windows-CAPI2/Operational",
      "Computer": "WIN-M5327EF98B9",
      "Security": {
        "#attributes": {
          "UserID": "S-1-5-21-1223297778-3299746493-1462173606-500"
        }
      }
    },
    "UserData": {
      "CryptRetrieveObjectByUrlWire": {
        "URL": {
          "#attributes": {
            "scheme": "http"
          },
          "#text": "http://crl.microsoft.com/pki/crl/products/microsoftrootcert.crl"
        },
        "Object": {
          "#attributes": {
            "type": "CONTEXT_OID_CRL",
            "constant": "2"
          }
        },
        "Timeout": "PT9.984S",
        "Flags": {
          "#attributes": {
            "value": "202005",
            "CRYPT_RETRIEVE_MULTIPLE_OBJECTS": "true",
            "CRYPT_WIRE_ONLY_RETRIEVAL": "true",
            "CRYPT_LDAP_SCOPE_BASE_ONLY_RETRIEVAL": "true",
            "CRYPT_PROXY_CACHE_RETRIEVAL": "true"
          }
        },
        "AuxInfo": {
          "#attributes": {
            "maxUrlRetrievalByteCount": "104857600",
            "fProxyCacheRetrieval": "true"
          }
        },
        "AdditionalInfo": {
          "NetworkConnectivityStatus": {
            "#attributes": {
              "value": "1",
              "_SENSAPI_NETWORK_ALIVE_LAN": "true"
            }
          },
          "Action": {
            "#attributes": {
              "name": "NoProxy"
            }
          },
          "Action_1": {
            "#attributes": {
              "name": "NoProxy"
            }
          },
          "HTTPRequestHeadersInfo": {
            "Header": "Connection: Keep-Alive",
            "Header_1": "GET /pki/crl/products/microsoftrootcert.crl HTTP/1.1",
            "Header_2": "Accept: */*",
            "Header_3": "If-None-Match: \"ea9ee7b1bc43d21:0\"",
            "Header_4": "If-Modified-Since: Mon, 21 Nov 2016 06:01:26 GMT",
            "Header_5": "Cache-Control: max-age = 900",
            "Header_6": "User-Agent: Microsoft-CryptoAPI/6.1"
          },
          "HTTPResponseHeadersInfo": {
            "Header": "x-ms-blob-type: BlockBlob",
            "Header_1": "HTTP/1.1 200 OK",
            "Header_2": "Connection: keep-alive",
            "Header_3": "Date: Thu, 18 May 2017 11:37:58 GMT",
            "Header_4": "Content-Length: 813",
            "Header_5": "Content-Type: application/pkix-crl",
            "Header_6": "Last-Modified: Tue, 02 May 2017 22:24:24 GMT",
            "Header_7": "ETag: 0x8D491A9FD112A27",
            "Header_8": "Server: Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0",
            "Header_9": "x-ms-request-id: 477c132d-0001-0045-443b-c49ae1000000",
            "Header_10": "x-ms-version: 2009-09-19",
            "Header_11": "x-ms-lease-status: unlocked"
          }
        },
        "CacheInfo": {
          "#attributes": {
            "lastSyncTime": "2017-05-19T02:02:36.203Z"
          },
          "URLCacheResponseInfo": {
            "#attributes": {
              "responseType": "CRYPTNET_URL_CACHE_RESPONSE_HTTP",
              "lastModifiedTime": "2017-05-02T22:24:24Z"
            }
          }
        },
        "RetrievedObjects": {
          "CertificateRevocationList": {
            "#attributes": {
              "fileRef": "0986764ED95D3C77F3F1AD8340EDD2F36C3BF8E7.crl",
              "issuerName": "Microsoft Root Certificate Authority"
            }
          }
        },
        "EventAuxInfo": {
          "#attributes": {
            "ProcessName": "Setup.exe"
          }
        },
        "CorrelationAuxInfo": {
          "#attributes": {
            "TaskId": "{74E4CD40-C966-49F5-B50A-032DFFEE57CA}",
            "SeqNumber": "7"
          }
        },
        "Result": {
          "#attributes": {
            "value": "0"
          }
        }
      }
    }
  }
}
